src/transformers/NestedLoopTransformation.cc

```diff
@@ -0,0 +1,342 @@
+/*
+ * Copyright (c) 2023, Martin Blicha <martin.blicha@gmail.com>
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include "NestedLoopTransformation.h"
+
+#include "QuantifierElimination.h"
+#include "TransformationUtils.h"
+#include "graph/ChcGraph.h"
+#include "utils/SmtSolver.h"
+
+std::unique_ptr<NestedLoopTransformation::WitnessBackTranslator>
+NestedLoopTransformation::transform(ChcDirectedGraph & graph) {
+    FindFirstLoop loopDetector;
+    auto vertices = graph.getVertices();
+    auto originalGraph = graph;
+    std::vector<EId> loop = loopDetector.detectLoop(graph);
+    std::vector<SymRef> simplifiedNodes;
+    LocationVarMap locationVars;
+    PositionVarMap argVars;
+    while (loop.size() > 1) {
+        simplifiedNodes.push_back(simplifyLoop(graph, loop, locationVars, argVars));
+        loop = loopDetector.detectLoop(graph);
+    }
+    return std::make_unique<NestedLoopTransformation::WitnessBackTranslator>(
+        std::move(originalGraph), graph, std::move(locationVars), std::move(argVars), simplifiedNodes);
+}
+
+SymRef NestedLoopTransformation::simplifyLoop(ChcDirectedGraph & graph, std::vector<EId> loop,
+                                              LocationVarMap & locationVars, PositionVarMap & argVars) {
+    Logic & logic = graph.getLogic();
+    TimeMachine timeMachine(logic);
+    TermUtils utils(logic);
+    std::vector<SymRef> vertices;
+    assert(graph.getSource(loop[0]) == graph.getTarget(loop[loop.size() - 1]));
+    assert(loop.size() >= 2);
+    for (int i = loop.size() - 1; i >= 0; i--) {
+        vertices.push_back(graph.getSource(loop[i]));
+    }
+    // KB: creation of the location variables for nodes and argument variables (params of predicates)
+    std::vector<EId> allEdges = graph.getEdges();
```

Comment:
Move this to the place where it is actually used.

Reply:
Done